node/core/provisioner/src/lib.rs

```diff
@@ -503,6 +562,108 @@ async fn select_candidates(
 		}
 	}
 
+	Ok(selected_candidates)
+}
+
+/// Requests backable candidates from Prospective Parachains subsystem
+/// based on core states.
+///
+/// Should be called when prospective parachains are enabled.
+async fn request_backable_candidates(
+	availability_cores: &[CoreState],
+	bitfields: &[SignedAvailabilityBitfield],
+	relay_parent: Hash,
+	sender: &mut impl overseer::ProvisionerSenderTrait,
+) -> Result<Vec<CandidateHash>, Error> {
+	let block_number = get_block_number_under_construction(relay_parent, sender).await?;
+
+	// Wrapped sender is shared among concurrent prospective parachains requests.
+	let wrapped_sender = Mutex::new(sender);
+	let mut selected_candidates = Vec::with_capacity(availability_cores.len());
+
+	for (core_idx, core) in availability_cores.iter().enumerate() {
+		let (para_id, required_path) = match core {
+			CoreState::Scheduled(scheduled_core) => {
+				// The core is free, pick the first eligible candidate from
+				// the fragment tree.
+				(scheduled_core.para_id, Vec::new())
+			},
+			CoreState::Occupied(occupied_core) => {
+				if bitfields_indicate_availability(core_idx, bitfields, &occupied_core.availability)
+				{
+					if let Some(ref scheduled_core) = occupied_core.next_up_on_available {
+						// The candidate occupying the core is available, choose its
+						// child in the fragment tree.
+						//
+						// TODO [now]: doesn't work for parathreads.
+						//
+						// ?? match scheduled para id == candidate_para_id?
+						(scheduled_core.para_id, vec![occupied_core.candidate_hash])
```

Comment:
@rphmeier if the next para is not the same as the one occupying the core, `required_path` should be empty, right?
Otherwise we will try to walk the fragment tree of `next_para_id` with candidate hash from `curr_para_id`.
Thus prospective parachains will return `None` with this code if parathreads are also enabled.

Reply:
> if the next para is not the same as the one occupying the core, required_path should be empty, right?

Yeah. Please change the doc comment to say something like "TODO: We lean hard on the assumption that cores are fixed to specific parachains within a session. This breaks for parathreads" and we should link to some issue about it. Scheduler runtime APIs will need to be adapted for parathreads anyway.